### DISCUSSION TITLE
tx cost issue during submission to blockfrost: New issue

### DISCUSSION THREAD

**Discussion by reitcircles** [Python-Cardano/pycardano#307] [General]
Failed to submit transaction. Error code: 400. Error message: "transaction submit error ShelleyTxValidationError ShelleyBasedEraBabbage (ApplyTxError [UtxowFailure (UtxoFailure (FromAlonzoUtxoFail (FeeTooSmallUTxO (Coin 299490) (Coin 295088))))])"

Tried to change the minimum fees but it fails to assign to the signed tx. 
So basically now the library of pycardano is not compatible with tx submission to Blockfrost

----------------------for the following tx---------------------------

############### Transaction created ###############
{
  'auxiliary_data': AuxiliaryData(data=AlonzoMetadata(metadata={721: {'3026aeee3737f5010ca04d9cb9938efead6178d1db02135feea9cfaa': {AssetName(b'(100)USER'): {'tokenType': 'role', 'description': 'Role NFT from reitcircles.com', 'image': 'QmNmrtWycJARzQu7qWwVwB4azurBmCR2oSp3d9dtLdusCQ'}}}}, native_scripts=None, plutus_scripts=None)),
  'transaction_body': {
    'auxiliary_data_hash': AuxiliaryDataHash(hex='e7842593f77a2742da7579fabe838e39c27d8fd9646766ea374b4e00455f3d9c'),
    'certificates': None,
    'collateral': [
      {
        'index': 0,
        'transaction_id': TransactionId(hex='d9039c95d1160cf6bdf49c80708fe08e043830761c7c41d159581a36cbf511d3'),
      },
    ],
    'collateral_return': {
      'address': addr_test1qzylgj0pzzm722x3vkp6nde7h5cuzt2m4g9c69n5dj0w5rqd74hhzeetur4qr9v9e2mce0uxlzqelhuy9dq2027dh78sgh34ld,
      'amount': {'coin': 1392385, 'multi_asset': {}},
      'datum': None,
      'datum_hash': None,
      'post_alonzo': False,
      'script': None,
    },
    'fee': 295088,
    'inputs': [
      {
        'index': 2,
        'transaction_id': TransactionId(hex='52de516ff479908b7573ea1cf5c5fca69d5bb7318e0cb92f93f04b15571b72fe'),
      },
    ],
    'mint': {ScriptHash(hex='3026aeee3737f5010ca04d9cb9938efead6178d1db02135feea9cfaa'): {AssetName(b'(222)REFERENCE'): 1, AssetName(b'(100)USER'): 1}},
    'network_id': None,
    'outputs': [
      {
        'address': addr_test1wrgzpd4vlelrt6pvg8fav4yc2v9zsqppz2w4tst8tuu4sgg20dmpg,
        'amount': {
          'coin': 25000000,
          'multi_asset': {ScriptHash(hex='3026aeee3737f5010ca04d9cb9938efead6178d1db02135feea9cfaa'): {AssetName(b'(222)REFERENCE'): 1}},
        },
        'datum': RoleDatum(roleName=b'general', subscriptionLevel=b'Tier1', expiryDate=1688064662, pubkeyhash=b'\x89\xf4I\xe1\x10\xb7\xe5(\xd1e\x83\xa9\xb7>\xbd1\xc1-[\xaa\x0b\x8d\x16tl\x9e\xea\x0c'),
        'datum_hash': None,
        'post_alonzo': False,
        'script': None,
      },
      {
        'address': addr_test1qrwcxmlx56njze57fjypt2f6q2ctpv2tvx2gf2mf866ajsasqj43gyduezd0crpju8l9ps4fuu94z0rp86t00c4m3nzs8yf0nd,
        'amount': {
          'coin': 25000000,
          'multi_asset': {ScriptHash(hex='3026aeee3737f5010ca04d9cb9938efead6178d1db02135feea9cfaa'): {AssetName(b'(100)USER'): 1}},
        },
        'datum': None,
        'datum_hash': None,
        'post_alonzo': False,
        'script': None,
      },
      {
        'address': addr_test1qzylgj0pzzm722x3vkp6nde7h5cuzt2m4g9c69n5dj0w5rqd74hhzeetur4qr9v9e2mce0uxlzqelhuy9dq2027dh78sgh34ld,
        'amount': {'coin': 344432877, 'multi_asset': {}},
        'datum': None,
        'datum_hash': None,
        'post_alonzo': False,
        'script': None,
      },
    ],
    'reference_inputs': None,
    'required_signers': [
      VerificationKeyHash(hex='89f449e110b7e528d16583a9b73ebd31c12d5baa0b8d16746c9eea0c'),
    ],
    'script_data_hash': ScriptDataHash(hex='1bbdc3fdf8aa38b1741525542ae8bd46df683023aa6a9d8008dc1db99cbf8ed5'),
    'total_collateral': 3607615,
    'ttl': 40856414,
    'update': None,
    'validity_start': 40845414,
    'withdraws': None,
  },
  'transaction_witness_set': {
    'bootstrap_witness': None,
    'native_scripts': None,
    'plutus_data': None,
    'plutus_v1_script': None,
    'plutus_v2_script': [
      b'Y\x08\x95\x01\x00\x002"##7J\x90\x00\x1b\xb1I\x8c\x8c\x8c\xcc\xcc\x00@\x18\x01I&\x000\x02S0\x01H\x81\x01\r\x00\x13&3W8\x92\x01\x14NameError: validator\x00I\x8c\x8c\x8c\x88\x94\xcc\xd5\xcd\x19\xb8\xf0\x02H\x81\x01\r\x00\x11\x0013\x00@\x02\x00\x12""2I\x8c\x8c\x94\xcc\xd5\xcd\x19)\x98\x00\xa4A\x01\x03\x00\x13&3W8\x92\x01\x16NameError: sig_present\x00I\x80\x04@\x04L\x99&23220\x01\x00\x12"S3Us\xe0\x04 \x02&f\x00`\x06j\xe8\x80\x08\xcc\x01\x00\x04\xd5\xd0\x80\x11\x11\x91\x92\x99\x9a\xb9\xa3#7\x1eddn\xb8\xcc\xc8\xc0\x04\x00H\x89L\xcdU\xcf\x80\x08\x991\x9a\xb9\xc4\x91\nIndexError\x00I\x85L\xcd\\\xd1\x9b\x87\x00$\x80\x00M]\x08\x00\x89\x99\x80\x18\x01\x99\xb8\x10\x02H\x00\x8d]\x10\x00\xa4\x00\xc6\xaa\xe7\x8d\xd5\x00\t\x92\x99\x80\nE\x01\x02\x00\x13&3W8\x92\x01\x11NameError: params\x00I\x80\x04\x00L\x94\xcc\x00R!\x01\x0c\x00\x13&3W8\x92\x01\x0cNameError: s\x00I\x80\x04\x00D\xc8\xc8\xc8\x89L\xcd\\\xd1\x9b\x8f\x00$\x89\x01\x03\x00\x11\x0013\x00@\x02\x00\x13$\xa2\x00 \x02\x00"d\x00 \x02\x00&DJfj\xe6\x8c\xdcx\x01$A\x01\x0c\x00\x11\x0013\x00P\x02\x00\x10\x01223##\x00\x10\x01"S3Us\xe0\x02)z\xc4\x135t\x06\x00f\xae\x84\x00L\xc0\x08\x00\x8d]\x10\x00\x91\xba\xe0\x017Xfd`\x02\x00$D\xa6f\xaa\xe7\xc0\x04L\x98\xcd\\\xe2I\nIndexError\x00I\x85L\xcd\\\xd1\x9b\x87\x00$\x80\x00M]\x08\x00\x89\x99\x80\x18\x01\x99\xb8\x10\x02H\x00\x8d]\x10\x00\xa4\x02\x06\xaa\xe7\x8d\xd5\x00\t\x91\x91\x99\x91\x80\x08\x00\x91\x12\x99\x9a\xab\x9f\x00\x112c5s\x89!\nIndexError\x00I\x85L\xcd\\\xd1\x9b\x87\x00$\x80\x00M]\x08\x00\x89\x99\x80\x18\x01\x99\xb8\x10\x02H\x00\x8d]\x10\x00\xa4\x00\x06\xaa\xe7\x8d\xd5\x00\t\x92\x99\x80\nE\x01\x00\x00\x13&3W8\x92\x01\x12NameError: context\x00I\x80\x04\x00@\x04\x00L\x8c\x88\x94\xcc\xd5\xcd\x19\xb8\xf0\x02H\x81\x01\x03\x00\x11\x0013\x00@\x02\x00\x13$\xa0\x00&F@\x04dfd\xa6`\x02\x91\x01\x01\x05\x00\x13&3W8\x92\x01!NameError: assert_minting_purpose\x00I\x80\x04\xc9L\xc0\x05"\x01\x01\x00\x00\x13&3W8\x92\x01\x12NameError: context\x00I\x80\x04\x00@\x04\xc9L\xcd\\\xd1\x91\x99\x99)\x98\x00\xa4A\x01\x08\x00\x13&3W8\x92\x01\x13NameError: has_utxo\x00I\x80\x04\xc8\xc8\xcc\xc8\xc0\x04\x00H\x89L\xcdU\xcf\x80\x08\x991\x9a\xb9\xc4\x91\nIndexError\x00I\x85L\xcd\\\xd1\x9b\x87\x00$\x80\x00M]\x08\x00\x89\x99\x80\x18\x01\x99\xb8\x10\x02H\x00\x8d]\x10\x00\xa4\x00\x06\xaa\xe7\x8d\xd5\x00\t\x92\x99\x80\nE\x01\x02\x00\x13&3W8\x92\x01\x11NameError: params\x00I\x80\x04\x00L\x94\xcc\x00R!\x01\x00\x00\x13&3W8\x92\x01\x12NameError: context\x00I\x80\x04\x00@\x04@\x04L\x98\xcd\\\xe1\x92I\x11UTxO not consumed\x00\x00\x14\x98\x00L\xcc\xc8\x88\x88\x94\xcc\xd5\xcd\x19\xb8\xf0\x02H\x81\x01\x00\x00\x11\x001S3W4f\xe3\xc0\t"\x01\x01\x0f\x00\x11\x00AS3W4f\xe3\xc0\t"\x01\x01\x0e\x00\x11\x00QS3W4f\xe3\xc0\t"\x01\x01\x02\x00\x11\x00a3\x00p\x02\x00\x10\x05\x00@\x03\x00##"%35sFn<\x00\x92!\x01\x08\x00\x11\x0013\x00@\x02\x00\x12"232S0\x01H\x81\x01\x04\x00\x13&3W8\x92\x01\x0eNameError: any\x00I\x80\x04\xc8\xcc\x8c\x8c\x00@\x04\x89L\xcdU\xcf\x80\x08\xa5\xeb@L\xd5\xd0\x18\x01\x9a\xba\x10\x013\x00 \x025t@\x02FFn\xbc\xc9L\xc0\x05"\x01\x01\t\x00\x13&3W8\x92\x01\x0fNameError: oref\x00I\x80\x04\xc8\xc8\xcc\xc8\xc0\x04\x00H\x89L\xcdU\xcf\x80\x08\x991\x9a\xb9\xc4\x91\nIndexError\x00I\x85L\xcd\\\xd1\x9b\x87\x00$\x80\x00M]\x08\x00\x89\x99\x80\x18\x01\x99\xb8\x10\x02H\x00\x8d]\x10\x00\xa4\x00\x06\xaa\xe7\x8d\xd5\x00\t\x92\x99\x80\nE\x01\x0b\x00\x13&3W8\x92\x01\x0cNameError: i\x00I\x80\x04\x00L\x88\x94\xcc\xd5\xcd\x19\xb8\xf0\x02H\x81\x01\x0b\x00\x11\x0013\x00P\x02\x00\x10\x01223##\x00\x10\x01"S3Us\xe0\x02&\xecRa3W@`\x06j\xe8@\x04\xcc\x00\x80\x08\xd5\xd1\x00\t\x00\t\xba\xc33#\x00\x10\x01"%35W>\x00"d\xc6j\xe7\x12A\nIndexError\x00I\x85L\xcd\\\xd1\x9b\x87\x00$\x80\x00M]\x08\x00\x89\x99\x80\x18\x01\x99\xb8\x10\x02H\x00\x8d]\x10\x00\xa4\x00\x06\xaa\xe7\x8d\xd5\x00\t\x91\x91\x99\x91\x80\x08\x00\x91\x12\x99\x9a\xab\x9f\x00\x112c5s\x89!\nIndexError\x00I\x85L\xcd\\\xd1\x9b\x87\x00$\x80\x00M]\x08\x00\x89\x99\x80\x18\x01\x99\xb8\x10\x02H\x00\x8d]\x10\x00\xa4\x00\x06\xaa\xe7\x8d\xd5\x00\t\x92\x99\x80\nE\x01\n\x00\x13&3W8\x92\x01\x12NameError: context\x00I\x80\x04\x00@\x04\x00@\x04\xc8\x00L\xc8\x88\x94\xcc\xd5\xcd\x19\xb8\xf0\x02H\x81\x01\n\x00\x11\x001S3W4f\xe3\xc0\t"\x01\x01\t\x00\x11\x00A3\x00P\x02\x00\x10\x03\x00##"%35sFn<\x00\x92!\x01\x05\x00\x11\x0013\x00@\x02\x00\x12#$\x98\xc8\xc9L\xcd\\\xd1\x92\x99\x80\nE\x01\x01\x00\x13&3W8\x92\x01\x15NameError: is_minting\x00I\x80\x04@\x04L\x98\xcd\\\xe1\x92I\x13not minting purpose\x00\x00\x14\x98\xc9L\xcd\\\xd1\x91\x9b\x87225W:n\xa8\x00L\x94\xcc\x00R!\x01\x07\x00\x13&3W8\x92\x01\x12NameError: purpose\x00I\x80\x04\x00L\x92\x00\x00\x01\x00\x11222"S3W4f\xe3\xc0\t"\x01\x01\x01\x00\x11\x0013\x00@\x02\x00\x13$\xa2\x00 \x02\x00"dddD\xa6f\xaeh\xcd\xc7\x80\x12E\x01\x01\x00\x11\x0013\x00@\x02\x00\x13$\xa0\x00 \x02\x00&FDJfj\xe6\x8c\xdcx\x01$A\x01\x07\x00\x11\x0013\x00@\x02\x00\x13##3#\x00\x10\x01"%35W>\x00"d\xc6j\xe7\x12A\nIndexError\x00I\x85L\xcd\\\xd1\x9b\x87\x00$\x80\x00M]\x08\x00\x89\x99\x80\x18\x01\x99\xb8\x10\x02H\x00\x8d]\x10\x00\xa4\x00F\xaa\xe7\x8d\xd5\x00\t\x92\x99\x80\nE\x01\x06\x00\x13&3W8\x92\x01\x12NameError: context\x00I\x80\x04\x00@\x04\xc8\x89L\xcd\\\xd1\x9b\x8f\x00$\x89\x01\x06\x00\x11\x0013\x00@\x02\x00\x10\x0222"S3W4f\xe3\xc0\t"\x01\x01\x04\x00\x11\x0013\x00@\x02\x00\x13"#3##\x00\x10\x01"%35W>\x00B\x00"f`\x06\x00f\xae\x88\x00\x8c\xc0\x10\x00M]\x08\x01\x11)\x99\xab\x9a\x00!J"\x00 \x04\x94\x00\x04\x00H\x80\x050\x01f\xd8y\x9f\xd8y\x9f\xd8y\x9fX R\xdeQo\xf4y\x90\x8bus\xea\x1c\xf5\xc5\xfc\xa6\x9d[\xb71\x8e\x0c\xb9/\x93\xf0K\x15W\x1br\xfe\xff\x02\xffI(100)USERN(222)REFERENCEX\x1c\x89\xf4I\xe1\x10\xb7\xe5(\xd1e\x83\xa9\xb7>\xbd1\xc1-[\xaa\x0b\x8d\x16tl\x9e\xea\x0c\xff\x00\x01',
    ],
    'redeemer': [
      {
        'data': {},
        'ex_units': {'mem': 1320, 'steps': 276120},
        'index': 0,
        'tag': {'__objclass__': <enum 'RedeemerTag'>, '_name_': 'MINT', '_value_': 1},
      },
    ],
    'vkey_witnesses': [
      {
        'signature': b'\xb7\xc7\xd0\xb1\x90\x7fgj\x95=$\xe6\xda\xf9%v\x01\x8f\x02\x01\xd6<w\xf2\xe3\x9b\xee\xc4\xf8!&\x89F0\xd8\xc8\xbd\x1b\x03\x81\x91\rdwC\xb7\n\xfe=\x0f]\xf1"=\xec\xeb0p\xcf\xa7\x93N\xe0\x0c',
        'vkey': {"type": "", "description": "", "cborHex": "58204727e389ae22a423d2194bac195cf65171d112f41333ad05e506e66a0609fedc"},
      },
      {
        'signature': b'\x0b\xbb\x9d\x9d\x80yc\xf06N\xfa\xcc\xfb\x07#\x06\xab\xed"\xf7\x97\xc7\xacB\x16\xf2\xdd\xc5*\x0b\xc4\x08~\x98EL\xd1\xce\xa7\xa1\xc4\xe2\xfd5\xe6\'\x0c\x13Cx\x05\xb7\xdd\xd1\xec<\x87\x12\x91\xb3\xde:\x1a\x0e',
        'vkey': {"type": "", "description": "", "cborHex": "58203d9067df3dace09cbd3ebf181cc3030a9b315936b781670b05c86bbb8b88bc47"},
      },
    ],
  },
  'valid': True,
}

**Comment by reitcircles** [comment]
So basically there has been an update at the blockfrost site that has broken the pycardano lib. I increased the cost in the pycardano (just multiplied the current calculation by 5) and submitted this tx: 7fc20e2d6b6a0505999aec22637e394472093bf3e17d7a5d22ab5996382c4e6a (preview network)

**Comment by reitcircles** [comment]
Pls advice how to best solve this problem ...probably some parameter updates has broken this.
I changed this line in txbuilder.py 
` self.fee = 5*self._estimate_fee()`

**Comment by nielstron** [comment]
I am not sure this is restricted to the blockfrost backend. I have experienced similar issues with pycardano using the Ogmios library. Sometimes I figured out that the issue was actually additional signing keys added to the transaction that had not been accounted for during fee calculation (and were not required either). So it might actually not be due to an incorrect calculation but due to incorrectly assembled transactions. It would be nice to be able to add a fixed additional fee though as a quick way to circumvent these issues.

Btw 5x might be significantly too much in additional fee. It could be completely sufficient to just add a constant of 100 - 1000 lovelace, did you check this?

**Comment by reitcircles** [comment]
No. It popped up suddenly without any changes on our side. So it cannot be
related to our code.

**Reply by nielstron** [reply]
Not in your code but in other code in pycardano (assuming you updated the pycardano library)
